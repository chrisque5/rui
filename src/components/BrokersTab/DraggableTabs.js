/* eslint max-classes-per-file: ["error", 2] */
/* eslint-disable react/no-multi-comp */
import { Tabs } from 'antd';
import { DragSource, DropTarget } from 'react-dnd';
import { Component, Children } from 'react';
import PropTypes from 'prop-types';

class TabNode extends Component {
  render() {
    const {
      connectDragSource,
      connectDropTarget,
      children,
    } = this.props;

    return connectDragSource(
      connectDropTarget(children),
    );
  }
}

const cardTarget = {
  drop(props, monitor) {
    const dragKey = monitor.getItem().index;
    const hoverKey = props.index;

    if (dragKey === hoverKey) {
      return;
    }

    props.moveTabNode(dragKey, hoverKey);
    // eslint-disable-next-line no-param-reassign
    monitor.getItem().index = hoverKey;
  },
};

const cardSource = {
  beginDrag(props) {
    return {
      id: props.id,
      index: props.index,
    };
  },
};

const WrapTabNode = DropTarget(
  'DND_NODE',
  cardTarget,
  (connect) => ({
    connectDropTarget: connect.dropTarget(),
  }),
)(
  DragSource(
    'DND_NODE',
    cardSource,
    (connect, monitor) => ({
      connectDragSource: connect.dragSource(),
      isDragging: monitor.isDragging(),
    }),
  )(TabNode),
);

class DraggableTabs extends Component {
  moveTabNode = (dragKey, hoverKey) => {
    const newOrder = this.props.tabOrder.slice();
    const { children } = this.props;

    Children.forEach(children, (c) => {
      if (newOrder.indexOf(c.key) === -1) {
        newOrder.push(c.key);
      }
    });

    const dragIndex = newOrder.indexOf(dragKey);
    const hoverIndex = newOrder.indexOf(hoverKey);

    newOrder.splice(dragIndex, 1);
    newOrder.splice(hoverIndex, 0, dragKey);
    this.props.orderChanged(newOrder);
  }

  renderTabBar = (props, DefaultTabBar) => (
    <DefaultTabBar {...props}>
      {(node) => (
        <WrapTabNode
          key={node.key}
          index={node.key}
          moveTabNode={this.moveTabNode}
        >
          {node}
        </WrapTabNode>
      )}
    </DefaultTabBar>
  );

  render() {
    const order = this.props.tabOrder;
    const { children } = this.props;

    const tabs = [];
    Children.forEach(children, (c) => {
      tabs.push(c);
    });

    const orderTabs = tabs.slice().sort((a, b) => {
      const orderA = order.indexOf(a.key);
      const orderB = order.indexOf(b.key);

      if (orderA !== -1 && orderB !== -1) {
        return orderA - orderB;
      }
      if (orderA !== -1) {
        return -1;
      }
      if (orderB !== -1) {
        return 1;
      }

      const ia = tabs.indexOf(a);
      const ib = tabs.indexOf(b);

      return ia - ib;
    });
    const { orderChanged, tabOrder, ...rest } = this.props;
    return (
      <Tabs
        renderTabBar={this.renderTabBar}
        {...rest}
      >
        {orderTabs}
      </Tabs>
    );
  }
}

export default DraggableTabs;

DraggableTabs.propTypes = {
  children: PropTypes.arrayOf(PropTypes.shape()).isRequired,
  orderChanged: PropTypes.func.isRequired,
  tabOrder: PropTypes.arrayOf(PropTypes.string).isRequired,
};

TabNode.propTypes = {
  children: PropTypes.shape().isRequired,
  connectDropTarget: PropTypes.func.isRequired,
  connectDragSource: PropTypes.func.isRequired,
};
