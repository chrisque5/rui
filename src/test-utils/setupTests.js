// The below part of the code is to fix the issue
// TypeError: window.matchMedia is not a function
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});
