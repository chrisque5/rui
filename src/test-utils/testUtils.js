/* eslint-disable import/no-extraneous-dependencies, react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Form } from 'antd';

const middlewares = [thunk];
const renderWithForm = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}><Form>{children}</Form></Provider>;

  return render(ui, { wrapper, ...renderOptions });
};

const renderWithProvider = (
  ui,
  {
    initialState,
    store = configureMockStore(middlewares)(initialState),
    ...renderOptions
  } = {},
) => {
  const wrapper = ({ children }) => <Provider store={store}>{children}</Provider>;

  return render(ui, { wrapper, ...renderOptions });
};
// re-export everything
export * from '@testing-library/react';
// override render method
export { renderWithForm, renderWithProvider };
