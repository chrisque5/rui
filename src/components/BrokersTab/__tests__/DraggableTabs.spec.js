import {
  render, fireEvent, waitFor, getByText,
} from 'test-utils/testUtils';
import { act } from 'react-dom/test-utils';
import { DndProvider } from 'react-dnd';
import { TestBackend } from 'react-dnd-test-backend';

import { Tabs } from 'antd';
import DraggableTabs from '../DraggableTabs';

const defaultProps = {
  hideAdd: true,
  onChange: jest.fn(),
  type: 'editable-card',
  onEdit: jest.fn(),
  orderChanged: jest.fn(),
  tabOrder: [],
  activeKey: '1',
};

describe('<DraggableTabs/>', () => {
  test('renders snapshot as expected', async () => {
    const { container } = render(
      <DndProvider backend={TestBackend}>
        <DraggableTabs {...defaultProps}>
          <Tabs.TabPane key="t1" tab="Tab1"><div>Tab 1</div></Tabs.TabPane>
          <Tabs.TabPane key="t2" tab="Tab2"><div>Tab 2</div></Tabs.TabPane>
        </DraggableTabs>
      </DndProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  test.skip('Drag and drop', async () => {
    act(() => {
      render(
        <DndProvider backend={TestBackend}>
          <DraggableTabs {...defaultProps}>
            <Tabs.TabPane key="t1" tab="Tab1"><div>Tab 1</div></Tabs.TabPane>
            <Tabs.TabPane key="t2" tab="Tab2"><div>Tab 2</div></Tabs.TabPane>
          </DraggableTabs>
        </DndProvider>,
      );
    });

    const dragSource = await waitFor(() => getByText(document.body, 'Tab1'));
    const dropTarget = await waitFor(() => getByText(document.body, 'Tab2'));

    fireEvent.dragStart(dragSource);
    fireEvent.dragOver(dropTarget);
    fireEvent.drop(dropTarget);
    fireEvent.dragEnd(dragSource);

    expect(defaultProps.orderChanged).toHaveBeenCalled();
  });
});
