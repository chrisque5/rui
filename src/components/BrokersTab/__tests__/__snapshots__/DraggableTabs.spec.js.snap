// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DraggableTabs/> renders snapshot as expected 1`] = `
<div>
  <div
    class="ant-tabs ant-tabs-top ant-tabs-editable ant-tabs-card ant-tabs-editable-card"
  >
    <div
      class="ant-tabs-nav"
      role="tablist"
    >
      <div
        class="ant-tabs-nav-wrap"
      >
        <div
          class="ant-tabs-nav-list"
          style="transform: translate(0px, 0px);"
        >
          <div
            class="ant-tabs-tab ant-tabs-tab-with-remove"
          >
            <div
              aria-controls="rc-tabs-test-panel-t1"
              aria-selected="false"
              class="ant-tabs-tab-btn"
              id="rc-tabs-test-tab-t1"
              role="tab"
              tabindex="0"
            >
              Tab1
            </div>
            <button
              aria-label="remove"
              class="ant-tabs-tab-remove"
              tabindex="0"
              type="button"
            >
              <span
                aria-label="close"
                class="anticon anticon-close"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="ant-tabs-tab ant-tabs-tab-with-remove"
          >
            <div
              aria-controls="rc-tabs-test-panel-t2"
              aria-selected="false"
              class="ant-tabs-tab-btn"
              id="rc-tabs-test-tab-t2"
              role="tab"
              tabindex="0"
            >
              Tab2
            </div>
            <button
              aria-label="remove"
              class="ant-tabs-tab-remove"
              tabindex="0"
              type="button"
            >
              <span
                aria-label="close"
                class="anticon anticon-close"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="ant-tabs-ink-bar ant-tabs-ink-bar-animated"
          />
        </div>
      </div>
      <div
        class="ant-tabs-nav-operations ant-tabs-nav-operations-hidden"
      >
        <button
          aria-controls="rc-tabs-test-more-popup"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-hidden="true"
          class="ant-tabs-nav-more"
          id="rc-tabs-test-more"
          style="visibility: hidden; order: 1;"
          tabindex="-1"
          type="button"
        >
          <span
            aria-label="ellipsis"
            class="anticon anticon-ellipsis"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="ellipsis"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="ant-tabs-content-holder"
    >
      <div
        class="ant-tabs-content ant-tabs-content-top"
      >
        <div
          aria-hidden="true"
          aria-labelledby="rc-tabs-test-tab-t1"
          class="ant-tabs-tabpane"
          id="rc-tabs-test-panel-t1"
          role="tabpanel"
          style="display: none;"
          tabindex="-1"
        />
        <div
          aria-hidden="true"
          aria-labelledby="rc-tabs-test-tab-t2"
          class="ant-tabs-tabpane"
          id="rc-tabs-test-panel-t2"
          role="tabpanel"
          style="display: none;"
          tabindex="-1"
        />
      </div>
    </div>
  </div>
</div>
`;
